File: src/main/java/aurocosh/divinefavor/common/item/talismans/spell/SpellTalismanPlaceBlock.java
package aurocosh.divinefavor.common.item.talismans.spell;

import aurocosh.divinefavor.common.item.talismans.spell.base.ItemSpellTalisman;
import aurocosh.divinefavor.common.item.talismans.spell.base.SpellOptions;
import aurocosh.divinefavor.common.item.talismans.spell.base.TalismanContext;
import aurocosh.divinefavor.common.spirit.base.ModSpirit;
import aurocosh.divinefavor.common.util.UtilBlock;
import aurocosh.divinefavor.common.util.UtilPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;

import java.util.EnumSet;

public class SpellTalismanPlaceBlock extends ItemSpellTalisman {
    public SpellTalismanPlaceBlock(String name, ModSpirit spirit, int favorCost, EnumSet<SpellOptions> options) {
        super(name, spirit, favorCost, options);
    }

    @Override
    protected boolean validate(TalismanContext context) {
        BlockPos pos = context.pos.offset(context.facing);
        if (!UtilBlock.isAirOrReplaceable(context.world, pos))
            return false;
        EnumHand otherHand = UtilPlayer.getOtherHand(context.hand);
        ItemStack heldItem = context.player.getHeldItem(otherHand);
        Item item = heldItem.getItem();
        return item instanceof ItemBlock;
    }

    @Override
    protected void performActionServer(TalismanContext context) {
        BlockPos pos = context.pos.offset(context.facing);

        EnumHand otherHand = UtilPlayer.getOtherHand(context.hand);
        ItemStack heldItem = context.player.getHeldItem(otherHand);
        UtilBlock.replaceBlock(context.player, context.world, pos, heldItem, otherHand);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/item/spell_talismans/SpellTalismanPlaceBlock.kt
package aurocosh.divinefavor.common.item.spell_talismans

import aurocosh.divinefavor.common.item.spell_talismans.base.ItemSpellTalisman
import aurocosh.divinefavor.common.item.spell_talismans.base.SpellOptions
import aurocosh.divinefavor.common.item.spell_talismans.context.TalismanContext
import aurocosh.divinefavor.common.lib.extensions.getOther
import aurocosh.divinefavor.common.spirit.base.ModSpirit
import aurocosh.divinefavor.common.util.UtilBlock
import net.minecraft.item.ItemBlock
import java.util.*

class SpellTalismanPlaceBlock(name: String, spirit: ModSpirit, favorCost: Int, options: EnumSet<SpellOptions>) : ItemSpellTalisman(name, spirit, favorCost, options) {

    override fun validate(context: TalismanContext): Boolean {
        val pos = context.pos.offset(context.facing)
        if (!UtilBlock.isAirOrReplaceable(context.world, pos))
            return false
        val heldItem = context.player.getHeldItem(context.hand.getOther())
        val item = heldItem.item
        return item is ItemBlock
    }

    override fun performActionServer(context: TalismanContext) {
        val pos = context.pos.offset(context.facing)

        val otherHand = context.hand.getOther()
        val heldItem = context.player.getHeldItem(otherHand)
        UtilBlock.replaceBlock(context.player, context.world, pos, heldItem, otherHand)
    }
}
